File: python/deepstrike/runtime/run_group.py
"""L1 (RunGroup) — a governance domain shared by N peer agent sessions of one logical run.

The kernel (execution vehicle) is ephemeral and torn down between stateless turns, so the cumulative
budget + membership that must span the whole group live outside any vehicle: in a ``GroupBudgetStore``.
Every store atomically holds capacity for each member and settles actual consumption. Per spec §2.5, only
*cumulative* budget is shared this way; instantaneous concurrency stays vehicle-scoped.

``InMemoryGroupBudgetStore`` provides process-local atomic reservations for one replica / tests.
"""

from __future__ import annotations

import asyncio
import uuid
from dataclasses import dataclass
from typing import Protocol, runtime_checkable


@dataclass
class GroupLedger:
    """Cumulative resources spent across a run group."""

    tokens_spent: int = 0
    subagents_spawned: int = 0
    # ③ loop-agent rounds completed across the group (seeds the pacing trap's max_rounds).
    rounds_completed: int = 0


@dataclass(frozen=True)
class GroupBudgetGrant:
    """Capacity granted on the axes this member actually requested."""

    tokens: int | None = None
    subagents: int | None = None
    rounds: int | None = None


@dataclass
class GroupMember:
    """A persona session that participated in the logical run (process-table lineage)."""

    session_id: str
    role: str | None = None
    # W-N5: what this member IS in the lineage — a "peer" persona (ReactiveSession.add_peer) vs a
    # "vehicle" session (run()/run_workflow envelopes, workflow-node children, loop iterations).
    # ``ReactiveSession.resume()`` rebuilds the peer set from "peer" members only, so DAG-in-Peer
    # usage can't resurrect phantom ``wf-node*`` personas. None (legacy) = unknown.
    kind: str | None = None  # "peer" | "vehicle"


@runtime_checkable
class GroupBudgetStore(Protocol):
    """Cumulative-budget + membership ledger shared by the members of a run group."""

    async def join(self, group_id: str, member: GroupMember) -> None:
        """Register a persona session as a member of the group (idempotent by session_id)."""
        ...

    async def members(self, group_id: str) -> list[GroupMember]:
        """All persona sessions of the logical run — the cross-invocation lineage (R2)."""
        ...


    async def reserve(
        self,
        group_id: str,
        *,
        member_id: str,
        limits: dict[str, int],
        requested: dict[str, int],
    ) -> "GroupBudgetReservation": ...

    async def settle(
        self,
        group_id: str,
        reservation_id: str,
        *,
        tokens: int = 0,
        subagents: int = 0,
        rounds: int = 0,
    ) -> None:
        """Idempotently replace a reservation with actual usage."""
        ...

    async def release(self, group_id: str, reservation_id: str) -> None:
        """Idempotently discard an unused reservation."""
        ...

class InMemoryGroupBudgetStore:
    """Process-local default store. One ledger + member set per group id."""

    def __init__(self) -> None:
        self._ledgers: dict[str, GroupLedger] = {}
        self._members: dict[str, dict[str, GroupMember]] = {}
        self._reservations: dict[str, dict[str, GroupBudgetReservation]] = {}
        self._locks: dict[str, asyncio.Lock] = {}

    async def read(self, group_id: str) -> GroupLedger:
        cur = self._ledgers.get(group_id)
        return GroupLedger(cur.tokens_spent, cur.subagents_spawned, cur.rounds_completed) if cur else GroupLedger()

    async def join(self, group_id: str, member: GroupMember) -> None:
        # Idempotent by session_id: the FIRST join wins. W-N5 relies on this — a persona joined as
        # "peer" must not be re-tagged "vehicle"
        # when its own turn's run() joins the same session id.
        self._members.setdefault(group_id, {}).setdefault(member.session_id, member)

    async def members(self, group_id: str) -> list[GroupMember]:
        return list(self._members.get(group_id, {}).values())

    async def reserve(
        self,
        group_id: str,
        *,
        member_id: str,
        limits: dict[str, int],
        requested: dict[str, int],
    ) -> "GroupBudgetReservation":
        async with self._locks.setdefault(group_id, asyncio.Lock()):
            settled = await self.read(group_id)
            held = GroupLedger()
            for reservation in self._reservations.get(group_id, {}).values():
                held.tokens_spent += reservation.granted.tokens or 0
                held.subagents_spawned += reservation.granted.subagents or 0
                held.rounds_completed += reservation.granted.rounds or 0
            ledger = GroupLedger(
                settled.tokens_spent + held.tokens_spent,
                settled.subagents_spawned + held.subagents_spawned,
                settled.rounds_completed + held.rounds_completed,
            )

            def grant(axis: str, used: int) -> int:
                wanted = max(0, requested.get(axis, 0))
                limit = limits.get(axis)
                return max(0, min(wanted, wanted if limit is None else limit - used))

            reservation = GroupBudgetReservation(
                id=str(uuid.uuid4()),
                group_id=group_id,
                member_id=member_id,
                granted=GroupBudgetGrant(
                    tokens=grant("tokens", ledger.tokens_spent) if "tokens" in requested else None,
                    subagents=(
                        grant("subagents", ledger.subagents_spawned)
                        if "subagents" in requested else None
                    ),
                    rounds=grant("rounds", ledger.rounds_completed) if "rounds" in requested else None,
                ),
            )
            self._reservations.setdefault(group_id, {})[reservation.id] = reservation
            return reservation

    async def settle(
        self,
        group_id: str,
        reservation_id: str,
        *,
        tokens: int = 0,
        subagents: int = 0,
        rounds: int = 0,
    ) -> None:
        async with self._locks.setdefault(group_id, asyncio.Lock()):
            reservations = self._reservations.get(group_id, {})
            if reservations.pop(reservation_id, None) is None:
                return
            if not reservations:
                self._reservations.pop(group_id, None)
            cur = self._ledgers.setdefault(group_id, GroupLedger())
            cur.tokens_spent += max(0, tokens)
            cur.subagents_spawned += max(0, subagents)
            cur.rounds_completed += max(0, rounds)

    async def release(self, group_id: str, reservation_id: str) -> None:
        async with self._locks.setdefault(group_id, asyncio.Lock()):
            reservations = self._reservations.get(group_id, {})
            reservations.pop(reservation_id, None)
            if not reservations:
                self._reservations.pop(group_id, None)


@dataclass(frozen=True)
class GroupBudgetReservation:
    id: str
    group_id: str
    member_id: str
    granted: GroupBudgetGrant


@dataclass
class RunGroup:
    """Binds a runner to a governance domain: a stable group id + the store its members share."""

    id: str
    budget_store: GroupBudgetStore


class GroupBudgetScope:
    """One member's reservation lifecycle."""

    def __init__(
        self,
        group: RunGroup,
        granted: GroupBudgetGrant,
        reservation_id: str,
    ) -> None:
        self._group = group
        self.granted = granted
        self.reservation_id = reservation_id
        self._closed = False

    @classmethod
    async def open(
        cls,
        group: RunGroup,
        member: GroupMember,
        *,
        limits: dict[str, int],
        requested: dict[str, int],
    ) -> "GroupBudgetScope":
        await group.budget_store.join(group.id, member)
        reservation = await group.budget_store.reserve(
            group.id,
            member_id=member.session_id,
            limits=limits,
            requested=requested,
        )
        return cls(group, reservation.granted, reservation.id)

    async def settle(self, *, tokens: int = 0, subagents: int = 0, rounds: int = 0) -> None:
        if self._closed:
            return
        await self._group.budget_store.settle(
            self._group.id,
            self.reservation_id,
            tokens=tokens,
            subagents=subagents,
            rounds=rounds,
        )
        self._closed = True

    async def release(self) -> None:
        if self._closed:
            return
        await self._group.budget_store.release(self._group.id, self.reservation_id)
        self._closed = True

    @property
    def closed(self) -> bool:
        return self._closed

File: node/src/runtime/run-group.ts
/**
 * L1 (RunGroup) — a governance domain shared by N peer agent sessions of one logical run.
 *
 * The kernel (execution vehicle) is ephemeral and torn down between stateless turns, so the
 * cumulative budget + membership that must span the whole group live outside any vehicle: in a
 * `GroupBudgetStore`. Every store atomically reserves capacity and settles actual consumption.
 * Per spec §2.5, only *cumulative* budget is shared this way; instantaneous concurrency stays
 * vehicle-scoped.
 *
 * `InMemoryGroupBudgetStore` provides process-local atomic reservations for one replica / tests.
 */
import { randomUUID } from "node:crypto"

/** Cumulative resources spent across a run group. */
export interface GroupLedger {
  /** ③ loop-agent rounds completed across the group (seeds the pacing trap's max_rounds). */
  roundsCompleted?: number
  /** Total tokens spent by all members. */
  tokensSpent: number
  /** Total sub-agents spawned by all members (running + completed). */
  subagentsSpawned: number
}

/** A member's contribution to charge back to the group ledger. */
// (rounds: ③ loop-agent — one per completed round.)
export interface GroupCharge {
  tokens?: number
  subagents?: number
  /** ③ loop-agent: completed rounds to add to the group's round count. */
  rounds?: number
}

export interface GroupBudgetRequest {
  /** Group-wide hard limits. An omitted axis is unbounded for this admission. */
  limits: GroupCharge
  /** Maximum capacity this member wants to hold for its lifetime. */
  requested: GroupCharge
}

export interface GroupBudgetReservation {
  id: string
  groupId: string
  memberId: string
  /** Capacity granted to this member. May be lower than requested when the group is nearly full. */
  granted: GroupCharge
}

/** A persona session that participated in the logical run (process-table lineage). */
export interface GroupMember {
  sessionId: string
  role?: string
  /** W-N5: what this member IS in the lineage — a `"peer"` persona (ReactiveSession.addPeer) vs a
   *  `"vehicle"` session (run()/runWorkflow envelopes, workflow-node children, loop iterations).
   *  `ReactiveSession.resume()` rebuilds the peer set from `"peer"` members only, so DAG-in-Peer
   *  usage can't resurrect phantom `wf-node*` personas. Absent (legacy) = unknown. */
  kind?: "peer" | "vehicle"
}

export interface GroupBudgetStore {
  /** Register a persona session as a member of the group (idempotent by sessionId). */
  join(groupId: string, member: GroupMember): void | Promise<void>
  /** All persona sessions of the logical run — the cross-invocation lineage (R2). */
  members(groupId: string): GroupMember[] | Promise<GroupMember[]>
  reserve(
    groupId: string,
    request: GroupBudgetRequest & { memberId: string },
  ): GroupBudgetReservation | Promise<GroupBudgetReservation>
  /** Idempotently replace a reservation with actual usage. */
  settle(groupId: string, reservationId: string, actual: GroupCharge): void | Promise<void>
  /** Idempotently discard an unused reservation. */
  release(groupId: string, reservationId: string): void | Promise<void>
}

/** Process-local default store. One ledger + member set per group id. */
export class InMemoryGroupBudgetStore implements GroupBudgetStore {
  private readonly ledgers = new Map<string, GroupLedger>()
  private readonly memberships = new Map<string, Map<string, GroupMember>>()
  private readonly reservations = new Map<string, Map<string, GroupBudgetReservation>>()

  read(groupId: string): GroupLedger {
    const ledger = this.ledgers.get(groupId)
    return ledger
      ? { ...ledger }
      : { tokensSpent: 0, subagentsSpawned: 0, roundsCompleted: 0 }
  }

  private applyCharge(groupId: string, delta: GroupCharge): void {
    const cur = this.read(groupId)
    this.ledgers.set(groupId, {
      tokensSpent: cur.tokensSpent + Math.max(0, delta.tokens ?? 0),
      subagentsSpawned: cur.subagentsSpawned + Math.max(0, delta.subagents ?? 0),
      roundsCompleted: (cur.roundsCompleted ?? 0) + Math.max(0, delta.rounds ?? 0),
    })
  }

  join(groupId: string, member: GroupMember): void {
    if (!this.memberships.has(groupId)) this.memberships.set(groupId, new Map())
    // First join wins (idempotent by sessionId). A persona registered as "peer" then re-joining
    // through its own run() as "vehicle" must not
    // lose its peer tag (W-N5), and the two stores must agree on which record survives.
    const members = this.memberships.get(groupId)!
    if (!members.has(member.sessionId)) members.set(member.sessionId, member)
  }

  members(groupId: string): GroupMember[] {
    return [...(this.memberships.get(groupId)?.values() ?? [])]
  }

  reserve(
    groupId: string,
    request: GroupBudgetRequest & { memberId: string },
  ): GroupBudgetReservation {
    const settled = this.read(groupId)
    const held = [...(this.reservations.get(groupId)?.values() ?? [])].reduce<GroupLedger>(
      (sum, reservation) => ({
        tokensSpent: sum.tokensSpent + (reservation.granted.tokens ?? 0),
        subagentsSpawned: sum.subagentsSpawned + (reservation.granted.subagents ?? 0),
        roundsCompleted: (sum.roundsCompleted ?? 0) + (reservation.granted.rounds ?? 0),
      }),
      { tokensSpent: 0, subagentsSpawned: 0, roundsCompleted: 0 },
    )
    const ledger: GroupLedger = {
      tokensSpent: settled.tokensSpent + held.tokensSpent,
      subagentsSpawned: settled.subagentsSpawned + held.subagentsSpawned,
      roundsCompleted: (settled.roundsCompleted ?? 0) + (held.roundsCompleted ?? 0),
    }
    const grant = (requested = 0, limit: number | undefined, used: number): number =>
      Math.max(0, Math.min(Math.max(0, requested), limit === undefined ? requested : limit - used))
    const reservation: GroupBudgetReservation = {
      id: randomUUID(),
      groupId,
      memberId: request.memberId,
      granted: {
        ...(request.requested.tokens !== undefined
          ? { tokens: grant(request.requested.tokens, request.limits.tokens, ledger.tokensSpent) }
          : {}),
        ...(request.requested.subagents !== undefined
          ? { subagents: grant(request.requested.subagents, request.limits.subagents, ledger.subagentsSpawned) }
          : {}),
        ...(request.requested.rounds !== undefined
          ? { rounds: grant(request.requested.rounds, request.limits.rounds, ledger.roundsCompleted ?? 0) }
          : {}),
      },
    }
    if (!this.reservations.has(groupId)) this.reservations.set(groupId, new Map())
    this.reservations.get(groupId)!.set(reservation.id, reservation)
    return reservation
  }

  settle(groupId: string, reservationId: string, actual: GroupCharge): void {
    const reservations = this.reservations.get(groupId)
    if (!reservations?.delete(reservationId)) return
    if (reservations.size === 0) this.reservations.delete(groupId)
    this.applyCharge(groupId, actual)
  }

  release(groupId: string, reservationId: string): void {
    const reservations = this.reservations.get(groupId)
    if (!reservations?.delete(reservationId)) return
    if (reservations.size === 0) this.reservations.delete(groupId)
  }
}

/** One member's reservation lifecycle. */
export class GroupBudgetScope {
  private closed = false

  private constructor(
    private readonly group: RunGroup,
    readonly granted: GroupCharge,
    readonly reservationId: string,
  ) {}

  static async open(
    group: RunGroup,
    member: GroupMember,
    request: GroupBudgetRequest,
  ): Promise<GroupBudgetScope> {
    await group.budgetStore.join(group.id, member)
    const reservation = await group.budgetStore.reserve(group.id, {
      ...request,
      memberId: member.sessionId,
    })
    return new GroupBudgetScope(group, reservation.granted, reservation.id)
  }

  async settle(actual: GroupCharge): Promise<void> {
    if (this.closed) return
    await this.group.budgetStore.settle(this.group.id, this.reservationId, actual)
    this.closed = true
  }

  get isClosed(): boolean {
    return this.closed
  }

  async release(): Promise<void> {
    if (this.closed) return
    await this.group.budgetStore.release(this.group.id, this.reservationId)
    this.closed = true
  }
}

/** Binds a runner to a governance domain: a stable group id + the store its members share. */
export interface RunGroup {
  /** Stable id for this logical run's governance domain; all members pass the same one. */
  id: string
  /** Shared cumulative-budget + membership store. */
  budgetStore: GroupBudgetStore
}
